contracts/src/TeleporterTokenDestination.sol

```diff
@@ -0,0 +1,85 @@
+// (c) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// SPDX-License-Identifier: Ecosystem
+
+pragma solidity 0.8.18;
+
+import {TeleporterMessageInput, TeleporterFeeInfo} from "@teleporter/ITeleporterMessenger.sol";
+import {TeleporterOwnerUpgradeable} from "@teleporter/upgrades/TeleporterOwnerUpgradeable.sol";
+import {IWarpMessenger} from
+    "@avalabs/subnet-evm-contracts@1.2.0/contracts/interfaces/IWarpMessenger.sol";
+import {IERC20} from "@openzeppelin/contracts@4.8.1/token/ERC20/IERC20.sol";
+
+abstract contract TeleporterTokenDestination is TeleporterOwnerUpgradeable {
+    address public constant WARP_PRECOMPILE_ADDRESS = 0x0200000000000000000000000000000000000005;
+    bytes32 public immutable blockchainID;
+
+    bytes32 public immutable sourceBlockchainID;
+    address public immutable teleporterTokenSourceAddress;
+    IERC20 public immutable token;
+
+    event SendTokens(bytes32 indexed teleporterMessageID, address indexed sender, uint256 amount);
+
+    constructor(
+        address teleporterRegistryAddress,
+        address teleporterManager,
+        bytes32 sourceBlockchainID_,
+        address teleporterTokenSourceAddress_,
+        address token_
+    ) TeleporterOwnerUpgradeable(teleporterRegistryAddress, teleporterManager) {
+        sourceBlockchainID = sourceBlockchainID_;
+        teleporterTokenSourceAddress = teleporterTokenSourceAddress_;
+        token = IERC20(token_);
+        blockchainID = IWarpMessenger(WARP_PRECOMPILE_ADDRESS).getBlockchainID();
+    }
+
+    function _send(
+        bytes32 destinationBlockchainID,
+        address destinationBridgeAddress,
+        address recipient,
+        uint256 amount,
+        uint256 primaryFee,
+        uint256 secondaryFee
+    ) internal virtual {
+        // increase bridge balance by amount
+        // send message to destinationBridgeAddress
+        bytes32 messageID = _sendTeleporterMessage(
+            TeleporterMessageInput({
+                destinationBlockchainID: destinationBlockchainID,
```

Comment:
I think this is supposed to be `destinationBlockchainID: sourceBlockchainID`? We should have the source blockchain as the outer ID, so we know where to send it first, and then we have `destinationBlockchainID` in the inner message so we know where to forward it.

Reply:
yup nice catch